src/traces/ohlc/hover.js

```diff
@@ -0,0 +1,111 @@
+/**
+* Copyright 2012-2018, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var Axes = require('../../plots/cartesian/axes');
+var Fx = require('../../components/fx');
+var Color = require('../../components/color');
+var fillHoverText = require('../scatter/fill_hover_text');
+
+var DIRSYMBOL = {
+    increasing: '▲',
+    decreasing: '▼'
+};
+
+module.exports = function hoverPoints(pointData, xval, yval, hovermode) {
+    var cd = pointData.cd;
+    var xa = pointData.xa;
+    var ya = pointData.ya;
+    var trace = cd[0].trace;
+    var t = cd[0].t;
+
+    var type = trace.type;
+    var minAttr = type === 'ohlc' ? 'l' : 'min';
+    var maxAttr = type === 'ohlc' ? 'h' : 'max';
+
+    // potentially shift xval for grouped candlesticks
+    var centerShift = t.bPos || 0;
+    var x0 = xval - centerShift;
+
+    // ohlc and candlestick call displayHalfWidth different things...
+    var displayHalfWidth = t.bdPos || t.tickLen;
+    var hoverHalfWidth = t.wHover;
+
+    // if two items are overlaying, let the narrowest one win
+    var pseudoDistance = Math.min(1, displayHalfWidth / Math.abs(xa.r2c(xa.range[1]) - xa.r2c(xa.range[0])));
+    var hoverPseudoDistance = pointData.maxHoverDistance - pseudoDistance;
+    var spikePseudoDistance = pointData.maxSpikeDistance - pseudoDistance;
+
+    function dx(di) {
+        var pos = di.pos - x0;
+        return Fx.inbox(pos - hoverHalfWidth, pos + hoverHalfWidth, hoverPseudoDistance);
+    }
+
+    function dy(di) {
+        return Fx.inbox(di[minAttr] - yval, di[maxAttr] - yval, hoverPseudoDistance);
+    }
+
+    function dxy(di) { return (dx(di) + dy(di)) / 2; }
+    var distfn = Fx.getDistanceFunction(hovermode, dx, dy, dxy);
+    Fx.getClosest(cd, distfn, pointData);
+
+    // skip the rest (for this trace) if we didn't find a close point
+    if(pointData.index === false) return [];
+
+    // we don't make a calcdata point if we're missing any piece (x/o/h/l/c)
+    // so we need to fix the index here to point to the data arrays
+    var cdIndex = pointData.index;
+    var di = cd[cdIndex];
+    var i = pointData.index = di.i;
+
+    var dir = di.dir;
+    var container = trace[dir];
+    var lc = container.line.color;
+
+    if(Color.opacity(lc) && container.line.width) pointData.color = lc;
+    else pointData.color = container.fillcolor;
+
+    pointData.x0 = xa.c2p(di.pos + centerShift - displayHalfWidth, true);
+    pointData.x1 = xa.c2p(di.pos + centerShift + displayHalfWidth, true);
+
+    pointData.xLabelVal = di.pos;
+
+    pointData.spikeDistance = dxy(di) * spikePseudoDistance / hoverPseudoDistance;
+    pointData.xSpike = xa.c2p(di.pos, true);
```

Comment:
From `candlestick_double-y-axis`, are we expected spike lines over this point to be black:

![peek 2018-04-16 15-31](https://user-images.githubusercontent.com/6675409/38830727-420a9206-418b-11e8-9fb7-4154936fefde.gif)


?

Reply:
> are we expecting spike lines over this point to be black

Yeah, that color fails the [`tinycolor.readability` test](https://github.com/plotly/plotly.js/blob/b938842854538d3b6b0bc3ff6eb2cb77964dc899/src/components/fx/hover.js#L1296) so we just pick a color that contrasts with the background.